### PR TITLE
soc: sam0: Kill off dts_fixup.h extras

### PR DESCRIPTION
**Pull request by Sizurka**
We can't kill the file entirely, as some samples/tests still use a few defines from it, but this brings it down the the "normal" level that other SoCs are at.

The first commit just moves everything to automatically generated DT defines which had no reason to be in dts_fixup.h.  The second part removes the SERCOM parts, which involves adding DTS aliases for the devices.  This allows the existing mapping for SERCOM to DTS definition to be preserved, though through a slightly modified name.

This was tested on some of the basic API tests (e.g. `tests/drivers/spi/spi_loopback` and other samples), but it shouldn't actually be changing any functionality.

### CODE REVIEW

File: soc/arm/atmel_sam0/samd21/soc.h [hunk 1]
```diff
@@ -68,4 +68,21 @@
 #define SOC_ATMEL_SAM0_APBB_FREQ_HZ SOC_ATMEL_SAM0_MCK_FREQ_HZ
 #define SOC_ATMEL_SAM0_APBC_FREQ_HZ SOC_ATMEL_SAM0_MCK_FREQ_HZ
 
+
+#define _SOC_ATMEL_SAM0_SERCOM_CHECK(dt_index, type, sercom_id) \
+	(IS_ENABLED(DT_ATMEL_SAM0_ ## type ## _ ## dt_index) && \
+	(DT_ATMEL_SAM0_ ## type ## _ ## dt_index ## _SERCOM_ID) == sercom_id)
+#define SOC_ATMEL_SAM0_SERCOM_IS_ENABLED(type, id) (	\
```
Comment (benpicco) [inline soc/arm/atmel_sam0/samd21/soc.h]: I think this should be moved to `atmel_sam0/common/soc_sercom.h`

Comment (Sizurka) [reply soc/arm/atmel_sam0/samd21/soc.h]: I've changed it to use direct DTS aliases, which removes the need for those macros entirely.  However, this means supporting SAME54 depends on #14897 to support the extra interrupts.

**Review by benpicco — APPROVED** [review]
It took me a bit to understand the significance of those aliases, but this indeed simplifies things considerably!

File: drivers/watchdog/wdt_sam0.c [hunk 2]
```diff
@@ -143,23 +141,23 @@ static int wdt_sam0_install_timeout(struct device *dev,
 				/* Ensure we have time for the early warning */
 				per += 1;
 			}
-			WDT_REGS->EWCTRL.bit.EWOFFSET = per - 1;
+			WDT->EWCTRL.bit.EWOFFSET = per - 1;
```
Comment (benpicco) [inline drivers/watchdog/wdt_sam0.c]: This should be `WDT->EWCTRL.bit.EWOFFSET = per - 1U;` now after 24d71431e9c9da493ce00eb284ac9b0b2cb1bdca

Comment (Sizurka) [reply drivers/watchdog/wdt_sam0.c]: ACK and rebased.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please split the first commit into separate commits for each peripheral type

**Comment by Sizurka** [comment]
Done, the non-SERCOM peripherals are split out now.

File: dts/arm/atmel/samd.dtsi [hunk 3]
```diff
@@ -33,6 +33,11 @@
 			<0x0080A048 0x4>;
 	};
 
+	aliases {
+		pinmux-a = &pinmux_a;
+		pinmux-b = &pinmux_b;
```
Comment (MaureenHelm) [inline dts/arm/atmel/samd.dtsi]: Why did you add aliases for pinmux and sercom, but not gpio? The generated dts instance indices aren't guaranteed to match the static hardware indices. Consider the case that gpio port a is disabled in dts, then gpio port b will be the first enabled instance in dts and get index 0.

File: drivers/watchdog/wdt_sam0.c [hunk 4]
```diff
@@ -12,8 +12,6 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(wdt_sam0);
 
-#define WDT_REGS ((Wdt *)DT_WDT_SAM0_BASE_ADDRESS)
```
Comment (MaureenHelm) [inline drivers/watchdog/wdt_sam0.c]: Please keep using the base address from dts rather than from ASF. We are doing this by convention for all the vendors

File: drivers/timer/sam0_rtc_timer.c [hunk 5]
```diff
@@ -20,7 +20,7 @@
 #include <sys_clock.h>
 
 /* RTC registers. */
-#define RTC0 ((RtcMode0 *) DT_RTC_SAM0_BASE_ADDRESS)
+#define RTC0 (&(RTC->MODE0))
```
Comment (MaureenHelm) [inline drivers/timer/sam0_rtc_timer.c]: Please keep using the base address from dts rather than from ASF. We are doing this by convention for all the vendors

File: dts/bindings/rtc/atmel,sam0-rtc.yaml [hunk 6]
```diff
@@ -22,10 +22,4 @@ properties:
       description: mmio register space
       generation: define
       category: required
-
-    clock-generator:
```
Comment (MaureenHelm) [inline dts/bindings/rtc/atmel,sam0-rtc.yaml]: Please keep this

File: drivers/usb/device/usb_dc_sam0.c [hunk 7]
```diff
@@ -21,8 +21,7 @@ LOG_MODULE_REGISTER(usb_dc_sam0);
 
 #define USB_SAM0_IN_EP 0x80
 
-#define REGS ((Usb *)DT_USB_DC_SAM0_BASE_ADDRESS)
```
Comment (MaureenHelm) [inline drivers/usb/device/usb_dc_sam0.c]: Please keep using the base address from dts rather than from ASF. We are doing this by convention for all the vendors

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> Done, the non-SERCOM peripherals are split out now.

Thank you! This makes it easier to review

File: dts/arm/atmel/samd.dtsi (hunk 3 above)
Comment (Sizurka) [reply dts/arm/atmel/samd.dtsi]: I only added them because they where the only ones used.  That is the current boards only reference the pinmux rather than the GPIO devices.  The GPIO driver itself is independent of the indexing.  I have no problem adding them for GPIO however.

That is, since the GPIO drivers are self contained, the indexing doesn't matter (i.e. it doesn't matter if index 0 refers to PORT A or PORT B, since it has the correct address from the DTS).  Basically, nothing in the GPIO instance has to reference anything looked up in absolute indices (e.g. clocking) so it's not required.

File: dts/bindings/rtc/atmel,sam0-rtc.yaml (hunk 6 above)
Comment (Sizurka) [reply dts/bindings/rtc/atmel,sam0-rtc.yaml]: Ok, I will put it back in, but right now (pending something like #15585) actually setting it would break the driver: it currently hard codes the clock frequency as the one from GCLK 0.

File: drivers/watchdog/wdt_sam0.c (hunk 4 above)
Comment (Sizurka) [reply drivers/watchdog/wdt_sam0.c]: ACK, changed.

File: drivers/timer/sam0_rtc_timer.c (hunk 5 above)
Comment (Sizurka) [reply drivers/timer/sam0_rtc_timer.c]: ACK, changed.

File: drivers/usb/device/usb_dc_sam0.c (hunk 7 above)
Comment (Sizurka) [reply drivers/usb/device/usb_dc_sam0.c]: ACK, changed.

**Comment by Sizurka** [comment]
Shippable failure appears to be because #15023 was merged and #14861 isn't yet merged (I can't add labels, or I'd have added a blocked label to #15023, #14867 should likely have it added).  So I don't think it's related.

**Review by MaureenHelm — APPROVED** [review]
